ingest/ingest_files.py

```diff
@@ -0,0 +1,125 @@
+"""Module for text, CSV, and TSV file types used in SCP ingest
+
+DESCRIPTION
+Module provides extract capabilities for text, CSV, and TSV file types
+
+"""
+import csv
+import mimetypes
+import os
+import re
+from itertools import islice
+
+
+class IngestFiles:
+    def __init__(self, file_path, allowed_file_types, is_MTX=False):
+        if not os.path.exists(file_path):
+            raise IOError(f"File '{file_path}' not found")
+        self.allowed_file_types = allowed_file_types
+        self.file_type, self.file = self.open_file(file_path)
+        # Keeps tracks of lines parsed
+        self.amount_of_lines = 0
+        self.is_MTX = is_MTX
+
+    def open_file(self, file_path):
+        """ Opens txt, csv, or tsv formatted files"""
+        open_file = open(file_path, encoding='utf-8-sig')
+        file_connections = {
+            # Remove BOM with encoding='utf-8-sig'
+            'text/csv': self.open_csv(open_file),
+            'text/plain': open_file,
+            'text/tab-separated-values': self.open_tsv(open_file),
+        }
+        # Check file type
+        file_type = self.get_file_type(file_path)[0]
+        # See if file type is allowed
+        if file_type in self.allowed_file_types:
+            # Return file object and type
+            return file_type, file_connections.get(file_type)
+        else:
+            raise ValueError(f"Unsupported file format. Allowed file types are: {' '.join(self.allowed_file_type)}")
+
+    # Inherited function
+    def extract(self):
+        """ Calls extract function for txt, csv, or tsv formatted files to
+            retrieve all contents from file.
+        """
+
+        file_type_extract_fns = {
+            'text/csv': self.extract_csv_or_tsv,
+            'text/plain': self.extract_txt,
+            'text/tab-separated-values': self.extract_csv_or_tsv,
+        }
+        return file_type_extract_fns.get(self.file_type)()
+
+    def split_line(self, line):
+        """Splits lines on file format-appropriate delimiters"""
+
+        if self.is_MTX:
+            return re.findall(r'[^,\t]+', line)
+        else:
+            return re.findall(r'[^,\s\t]+', line)
```

Comment:
Concerned that splitting on whitespace will cause issues where the ingest file has a .txt suffix but is actually a tab-delimited file where the value between the tab has spaces. example would be "BM_tsne.txt" file from SCP101 (A Census of Immune Cells) where values under the header "CELL_TYPE" looked like:
monocyte CD14+
developing B cells
CD4 T cells

Reply:
Resolved in 00ce5f8

Reply:
I'm not sure I see how commit 00ce5f8 would ensure that this example cluster file is parsed correctly (using "cat -vet" to renter tabs as ^I and end of line as $ so that spaces can be visualized)
NAME^IX^IY^ICLUSTER^ICELL_TYPE$
TYPE^Inumeric^Inumeric^Igroup^Igroup$
MantonBM1_HiSeq_1_AAACCTGAGCAGGTCA-1^I-1.8670639929757706^I2.9621075639874057^I4^ICD4 T cells$
MantonBM1_HiSeq_1_AAACCTGCACACTGCG-1^I1.0457630874000223^I12.07025739768945^I1^Imonocyte CD14+$

it seems the current code would end up with a list of 7 elements for first non-header line and a list of 6 elements for the second non-header line. If the file were properly tagged as a tsv file, each line would have five elements.

Reply:
I think this actually is backwards - if the file is MTX, then we _do_ want to allow splitting on space.  We have 10X MTX files that are space delimited.  Otherwise, we should not be splitting on spaces as we know we will be getting cell metadata that has values with spaces in it (e.g. ontology labels).

So I think it should be this instead:

```
if self.is_MTX:
            return re.findall(r'[^,\s\t]+', line)
        else:
            return re.findall(r'[^,\t]+', line
```)

Reply:
If the CLUSTER and CELL_TYPE columns were swapped, would we lose the CLUSTER annotation when the file is parsed as txt instead of tsv?

Reply:
No we wouldn't. A txt file is still comma or tab delimited.